src/writer.rs

```diff
@@ -46,22 +48,42 @@ where
     A: Sub<Output = A>,
     D: Dimension + RemoveAxis,
 {
-    let header = build_header(data, reference, A::DATA_TYPE);
+    let compression_level = Compression::fast();
+    let is_gz = is_gz_file(&header_path);
+    let (header, data_path) = prepare_header_and_paths(&header_path, data, reference, A::DATA_TYPE);
 
     // Need the transpose for fortran ordering used in nifti file format.
     let data = data.t();
 
-    let f = File::create(&path)?;
-    let mut writer = BufWriter::new(f);
-    if is_gz_file(&path) {
-        let mut e = GzEncoder::new(writer, Compression::fast());
-        write_header(&mut e, &header)?;
-        write_data(&mut e, &header, data)?;
-        let _ = e.finish()?; // Must use result
+    let header_file = File::create(header_path)?;
+    let mut header_writer = BufWriter::new(header_file);
```

Comment:
Since `GzEncoder` already uses a buffer for the gz encoding process, wouldn't it be best to leave `BufWriter` out if `is_gz`?

Reply:
TIL. Yes, sure.